massa-versioning-worker/src/versioning.rs

```diff
@@ -0,0 +1,1057 @@
+                self.state = state;
+            }
+        }
+    }
+
+    /// Given a corresponding VersioningInfo, check if state is coherent
+    /// it is coherent
+    ///   if state can be at this position (e.g. can it be at state "Started" according to given time range)
+    ///   if history is coherent with current state
+    /// Return false for state == MipState::Error
+    pub fn is_coherent_with(&self, versioning_info: &MipInfo) -> bool {
+        // Always return false for state Error or if history is empty
+        if matches!(&self.state, &MipState::Error) || self.history.is_empty() {
+            return false;
+        }
+
+        // safe to unwrap (already tested if empty or not)
+        let (initial_ts, initial_state_id) = self.history.first_key_value().unwrap();
+        if *initial_state_id != MipStateTypeId::Defined {
+            // self.history does not start with Defined -> (always) false
+            return false;
+        }
+
+        // Build a new VersionStateHistory from initial state, replaying the whole history
+        // but with given versioning info then compare
+        let mut vsh = MipStateHistory::new(initial_ts.now);
+        let mut advance_msg = Advance {
+            start_timestamp: versioning_info.start,
+            timeout: versioning_info.timeout,
+            threshold: Amount::zero(),
+            now: initial_ts.now,
+        };
+
+        for (adv, _state) in self.history.iter().skip(1) {
+            advance_msg.now = adv.now;
+            advance_msg.threshold = adv.threshold;
+            vsh.on_advance(&advance_msg);
+        }
+
+        // XXX: is there always full eq? Can we have slight variation here?
+        vsh == *self
+    }
+
+    /// Query state at given timestamp
+    pub fn state_at(
+        &self,
+        ts: MassaTime,
+        start: MassaTime,
+        timeout: MassaTime,
+    ) -> Result<MipStateTypeId, StateAtError> {
+        if self.history.is_empty() {
+            return Err(StateAtError::EmptyHistory);
+        }
+
+        // Optim: this avoids iterating over history (cheap to retrieve first item)
+        let first = self.history.first_key_value().unwrap(); // safe to unwrap
+        if ts < first.0.now {
+            // Before initial state
+            return Err(StateAtError::BeforeInitialState(first.1.clone(), ts));
+        }
+
+        // At this point, we are >= the first state in history
+        let mut lower_bound = None;
+        let mut higher_bound = None;
+        let mut is_after_last = false;
+
+        // Optim: this avoids iterating over history (cheap to retrieve first item)
+        let last = self.history.last_key_value().unwrap(); // safe to unwrap
+        if ts > last.0.now {
+            lower_bound = Some(last);
+            is_after_last = true;
+        }
+
+        if !is_after_last {
+            // We are in between two states in history, find bounds
+            for (adv, state_id) in self.history.iter() {
```

Comment:
If you think this is well-enough optimized, can you add your reasoning for using the .iter() instead of .range() (with e.g. a link to your benchmarks ;) )

Reply:
The benchmark was a way of exercising with Criterion; not sure if this is really relevant here as we will have only at max 4 entries in history (Defined, Stated, LockedIn, Active).
If the code is not clear or too long to read, I can switch to the range version. What do you think?

Reply:
> (Defined, Stated, LockedIn, Active).

This is what I'm unsure about. Doesn't the current code say the history can contain a lot of different Started states, with different thresholds?
Because if history contains 4 states, I agree with you the iterator is fine. If you can have history an history that contains (Defined, Started{a}, Started{b}, Started{c}, ..., LockedIn, Active), then it is a problem.

Reply:
This conversation will be resolved by only keeping one Started state in the history.